mkosi/distributions/opensuse.py

```diff
@@ -21,47 +21,131 @@ def filesystem(cls) -> str:
 
     @classmethod
     def install(cls, state: "MkosiState") -> None:
+        if state.config.base_image:
+            # We assume that the base image has been properly initialized and it
+            # contains all the metadata we need to install the additional
+            # packages.
+            return zypper_install(state, {*state.config.packages})
+
         return install_opensuse(state)
 
+    @classmethod
+    def remove_packages(cls, state: "MkosiState", packages: list[str]) -> None:
+        zypper_remove(state, packages)
+
     @staticmethod
     def initrd_path(kver: str) -> Path:
         return Path("boot") / f"initrd-{kver}"
 
 
+def invoke_zypper(state: MkosiState,
+                  global_opts: set[str],
+                  verb: str,
+                  verb_opts: set[str],
+                  *args: Any,
+                  with_apivfs: bool = False) -> None:
+    zypp_conf = state.root.joinpath("etc/zypp/zypp.conf")
+
+    cmdline = ["zypper", "--root", state.root, *global_opts, verb, *verb_opts, *args]
+
+    env=dict(ZYPP_CONF=zypp_conf)
+
+    if with_apivfs:
+        run_with_apivfs(state, cmdline, env=env)
+    else:
+        run(cmdline, env=env)
+
+
+def zypper_init(state: MkosiState) -> None:
+    state.root.joinpath("etc/zypp").mkdir(mode=0o755, parents=True, exist_ok=True)
+
+    # No matter if --root is used or not, zypper always considers its config
+    # files from the host environment. If we want to use our custom versions for
+    # the rootfs, we're left with two ways to specify them, depending on whether
+    # we need to customize a setting defined in zypp.conf or in zypper.conf. If
+    # it's in zypp.conf then the environment variable 'ZYPP_CONF' must be used
+    # (!) otherwise a custom zypper.conf can be specified with the global option
+    # '--config'.
+
+    zypp_conf = state.root.joinpath("etc/zypp/zypp.conf")
+
+    # For some reason zypper has no command line option to exclude the docs,
+    # this can only be configured via zypp.conf.
+    def yes_no(b: bool) -> str:
+        return "yes" if b else "no"
+
+    zypp_conf_content = f"""\
+    [main]
+    solver.onlyRequires = yes
+    rpm.install.excludedocs = {yes_no(state.config.with_docs)}
+    """
+
+    zypp_conf.write_text(zypp_conf_content)
+
+
+def zypper_addrepo(state: "MkosiState", url: str, name: str, caching: bool = False) -> None:
+    invoke_zypper(state, [], "addrepo", ["-c", "-k" if caching else "-K"], url, name)
+
+
+def zypper_removerepo(state: "MkosiState", repo: str) -> None:
+    invoke_zypper(state, [], "removerepo", [], repo)
+
+
+def zypper_modifyrepo(state: "MkosiState", repo: str, caching: bool = None) -> None:
```

Comment:
```suggestion
def zypper_modifyrepo(state: "MkosiState", repo: str, caching: bool = False) -> None:
```

Reply:
Nope, if caching is not specified we shouldn't try to change the caching setting for the repo. What you suggest would  disable caching if `caching` was not specified.

Reply:
This doesn't make sense. First of all, the typing as it stands is wrong. `None` is not a `bool`, so mypy will complain about this. You would need to type this as `Optional[bool]` if you want `None` to be a valid value. But using `False` here has no measurable difference, since the truthiness of `None` is false and using it in a bool context (as done below) will use it in this way, e.g.
```
>>> caching = None
>>> ["-k" if caching else "-K"]
['-K']
>>> caching = False
>>> ["-k" if caching else "-K"]
['-K']
```

Reply:
My initial intention was to anticipate the fact that `zypper_modifyrepo()` should be able to handle multiple repository properties, not just one as it's the case now with `caching`.

Hence if for example the helper has support for setting/unsetting properties `A` and `B`, calling `zypper_modifyrepo(state, repo, B=False)` should never try to change `A` property. To make this possible we need to make distinction between `A` set to a value and `A` not passed at all.

But given the fact that only one property is supported at the moment, making `caching` optional doesn't really make sense. So for now I changed the code to make `caching` parameter mandatory.

Reply:
Ok, but the code for this would need to look differently, because the above 
```python
["-k" if caching else "-K"]
```
will always evaluate to either `-k` or `-K` whatever is passed in (as long as it support the `__bool__` protocol). For something agnostic the code would need to look something like this
```python
def zypper_modifyrepo(state: "MkosiState", repo: str, caching: Optiona[bool] = None) -> None:
    zypper_args = []
    if caching is not None:
        zypper_args = ["-k" if caching else "-K"]
    invoke_zypper(state, [], "modifyrepo", *zypper_args, repo)
```

Reply:
Correct, that was my initial implementation but somehow screwed it up  by the commit that introduced `invoke_zypper()`.

But as I said, I finally make `caching` parameter mandatory for now.